src/completions/BufferAll.ts

```diff
@@ -53,17 +53,19 @@ export class BufferAllCompletionSource extends Completions.CompletionSourceFuse
     }
 
     private async updateOptions(exstr?: string) {
-        const tabs: browser.tabs.Tab[] = await Messaging.message(
+        const tabsPromise: Promise<browser.tabs.Tab[]> = Messaging.message(
             "commandline_background",
             "allWindowTabs",
         )
-        const windows = (await browserBg.windows.getAll()).reduce(
-            (acc, win) => {
+        const windowsPromise: Promise<{
+            [tabId: number]: browser.windows.Window
+        }> = browserBg.windows.getAll().then(windows =>
+            windows.reduce((acc, win) => {
                 acc[win.id] = win
                 return acc
-            },
-            {},
+            }, {}),
```

Comment:
> `            const response: {[windowId: number]: browser.windows.Window} = {}`

Code style question - I'm surprised that you put the type declaration on the variable here instead of as the return type of the function. Does Typescript style prefer leaving return types up to inference?

```typescript
    private async getWindows(): Promise<{[windowId: number]: browser.windows.Window}> {
        const windows = await browserBg.windows.getAll()
        const response = {}
        windows.forEach(win => response[win.id] = win)
        return response
    }
```

> With the runtime-dispatch library in v2 this is less bad. Perhaps you could take a look at that (v2 branch src/dispatch-test.ts and src/runtime-dispatch.ts) and see if it fits your requirements?

Will look at it, thanks.

Reply:
I put the declaration there because as I recall typescript isn't smart enough to back apply the signature of the function to the variable response, but it can use the type of response as the signature of the function. So this way I only have to declare it once and I don't have to wrap the type in a promise.

If typescript is smart enough to apply the function signature back up to `response` then I don't really mind where it goes, but I do slightly prefer that the type information isn't repeated unnecessarily.

Reply:
Turns out it doesn't matter because the type on `response` apparently doesn't do anything either way, even if we do both:
```typescript
    private async getWindows(): Promise<{[windowId: number]: browser.windows.Window}> {
        const windows = await browserBg.windows.getAll()
        const response: {[windowId: number]: browser.windows.Window} = {}
        windows.forEach(win => response[win.id] = win)
        response["foo bar!"] = "baz"
        return response
    }
```
I keep forgetting that this is an inference layer with hints papered over javascript, not a compiler that needs every single thing to be fully typed. Might be some crazy implicit conversions going on here too; I wouldn't be entirely surprised if you told me that a `{[number]:T}` is equivalent to `{[string]:T}`. This explains some of the issues I've run into. >_< Thanks!

Reply:
I think whenever you use the `obj[x]` notation you bypass type safety. The hints in this case are for whoever calls `getWindows()` and for documentation.

I don't know if you know, but keys to js objects are always stringified. If you want more type safety within the function you can use a Map<number, browser.windows.Window> instead of an object.